### PR TITLE
fix(telemetry): metric tags to lowercase

### PR DESCRIPTION
**Pull request by avara1986**
Agent transforms Telemetry Metrics to lower case but It's better if the library transform those tags first because telemetry metrics ID is the namespace + name + tags. If we make a mistake and add two tags, one of them uppercase and the other one lowercase, it counts as 2 different metrics

## Checklist
- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Review by smola — APPROVED** [review]

File: ddtrace/internal/telemetry/metrics.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def to_dict(self):
             "type": self.metric_type,
             "common": self.is_common_to_all_tracers,
             "points": self._points,
-            "tags": ["%s:%s" % (k, v) for k, v in self._tags.items()],
+            "tags": ["%s:%s" % (str(k).lower(), str(v).lower()) for k, v in self._tags.items()],
```
[the remark this answers is not in the document]
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py:84]: other developers who want to use `metric_worker.add_count_metric...` they wouldn't probably know the agent transforms tags to lower case, I think its a quality of life improvement :stuck_out_tongue: what do you think @smola ?

**Review by gnufede — APPROVED** [review]

**Review by gnufede — DISMISSED** [review]

File: ddtrace/internal/telemetry/metrics.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py]: I moved the validation to this method:
https://github.com/DataDog/dd-trace-py/pull/5656/files#diff-55020dfcfb1ffbd9c043ed441e8d94cffa8ae830d9e9da89a8509d5c5a1c0a83R75

I was thinking about your suggestion and I recall one thing about telemetry metrics ID: the identifier of a metric is the namespace + name + tags. If we make a mistake and add two tags, one of them uppercase and the other one lowercase, it counts as 2 different metrics.

let see the behavior of this test:
https://github.com/DataDog/dd-trace-py/pull/5656/files#diff-78d991117775c35ac2fe5c4062c4318237c61c3dfe095af7766a0750ed1ab82dR174

if we add two metrics:
```python
        telemetry_writer.add_count_metric(TELEMETRY_NAMESPACE_TAG_TRACER, "test-metric", 1, {"a": "B"})
        telemetry_writer.add_count_metric(TELEMETRY_NAMESPACE_TAG_TRACER, "test-metric", 1, {"A": "b"})
```
the result is one metric:
```
{
                "common": True,
                "metric": "test-metric",
                "points": [[1642544540, 2.0]],
                "tags": ["a:b"],
                "type": "count",
}
```
if we don't validate the lowercase, the result would be:

```
{
                "common": True,
                "metric": "test-metric",
                "points": [[1642544540, 1.0]],
                "tags": ["a:B"],
                "type": "count",
},
{
                "common": True,
                "metric": "test-metric",
                "points": [[1642544540, 1.0]],
                "tags": ["A:b"],
                "type": "count",
}
```

**Review by gnufede — APPROVED** [review]

**Review by ZStriker19 — DISMISSED** [review]

**Review by gnufede — APPROVED** [review]

**Review by christophe-papazian — APPROVED** [review]

**Comment by avara1986** [comment]
> I don't think I understand what is fixed. The updated tests do not have any change to the telemetry metrics id. Is there a regression test you can add?

yeah, I updated those test to validate this feature:
https://github.com/DataDog/dd-trace-py/pull/5656/files#diff-78d991117775c35ac2fe5c4062c4318237c61c3dfe095af7766a0750ed1ab82dR123

https://github.com/DataDog/dd-trace-py/pull/5656/files#diff-78d991117775c35ac2fe5c4062c4318237c61c3dfe095af7766a0750ed1ab82dR123

**Comment by majorgreys** [comment]
> telemetry metrics ID is the namespace + name + tags.

Where do we see this problem? Is this id being generated in the Agent or backend? What's the impact of having two different metrics IDs because of an upper and lower case?

**Comment by avara1986** [comment]
> > telemetry metrics ID is the namespace + name + tags.
> 
> Where do we see this problem? Is this id being generated in the Agent or backend? What's the impact of having two different metrics IDs because of an upper and lower case?

I'm not sure if the metrics data would be valid in backend services if we report two metrics at the same time with same ID. Two important points of this refactor is:
- Every library should have the same behavior when reports Telemetry Metric (https://github.com/DataDog/system-tests/pull/1111)
- If we want to report metrics in a future in a agentless applications, we will have a problem

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**